http2-flowcontrol-sim/src/Connection.java

```diff
@@ -0,0 +1,119 @@
+import java.util.ArrayList;
+import java.util.HashMap;
+
+public class Connection extends SimulationListener implements ClockListener {
+
+  private HashMap<Integer, Stream> streams;
+  private int initialWindow;
+  private int availableWindow;
+  private int publishedWindow;
+  private SimulationClock controller;
+  private int bandwidthAvailable;
+  private int latency;
+  private int bandwidth;
+
+  public Connection(SimulationClock clock, int initWindow, int latency, int bandwidth) {
+    streams = new HashMap<Integer, Stream>();
+    controller = clock;
+    controller.addListener(this);
+    initialWindow = initWindow;
+    availableWindow = initWindow;
+    this.latency = latency;
+    this.bandwidth = bandwidth;
+    bandwidthAvailable = this.bandwidth;
+  }
+
+  @Override
+  public void tick(int time) {
+    bandwidthAvailable = bandwidth;
+    refillWindow();
+    doTickActions(controller);
+  }
+
+  public int intialWindow() {
+    return initialWindow;
+  }
+
+  public int availableWindow() {
+    return availableWindow;
+  }
+
+  public int publishedWindow() {
+    return publishedWindow;
+  }
+
+  public int bandwidthAvailable() {
+    return bandwidthAvailable;
+  }
+
+  public int latency() {
+    return latency;
+  }
+
+  public HashMap<Integer, Stream> streams() {
+    return streams;
+  }
+
+  /**
+   * Increment the connection-wide initial window. Does not update the individual stream windows
+   *
+   * @param timeStep
+   * @param newWindow
+   */
+  public void initialWindow(int timeStep, int newWindow) {
+    int oldWindow = initialWindow;
+    initialWindow = Math.max(0, newWindow);
+    availableWindow += (initialWindow - oldWindow);
```

Comment:
If the previous window is really big, and then it is shrunk down to zero, won't this be negative?  Does your code handle this case properly?